pkg/registry/apis/secret/service/consolidation.go

```diff
@@ -0,0 +1,76 @@
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/grafana-app-sdk/logging"
+	"github.com/grafana/grafana/pkg/registry/apis/secret/contracts"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type ConsolidationService struct {
+	tracer                    trace.Tracer
+	globalDataKeyStore        contracts.GlobalDataKeyStorage
+	encryptedValueStore       contracts.EncryptedValueStorage
+	globalEncryptedValueStore contracts.GlobalEncryptedValueStorage
+	encryptionManager         contracts.EncryptionManager
+}
+
+func ProvideConsolidationService(
+	tracer trace.Tracer,
+	globalDataKeyStore contracts.GlobalDataKeyStorage,
+	encryptedValueStore contracts.EncryptedValueStorage,
+	globalEncryptedValueStore contracts.GlobalEncryptedValueStorage,
+	encryptionManager contracts.EncryptionManager,
+) contracts.ConsolidationService {
+	return &ConsolidationService{
+		tracer:                    tracer,
+		globalDataKeyStore:        globalDataKeyStore,
+		encryptedValueStore:       encryptedValueStore,
+		globalEncryptedValueStore: globalEncryptedValueStore,
+		encryptionManager:         encryptionManager,
+	}
+}
+
+func (s *ConsolidationService) Consolidate(ctx context.Context) error {
+	// Disable all active data keys.
+	// This will ensure that no new data can be encrypted with the old keys.
+	err := s.globalDataKeyStore.DisableAllDataKeys(ctx)
+	if err != nil {
+		return fmt.Errorf("disabling all data keys: %w", err)
+	}
+
+	// List all encrypted values.
+	encryptedValues, err := s.globalEncryptedValueStore.ListAll(ctx, contracts.ListOpts{}, nil)
```

Comment:
So here because we are not filtering by <= timestamp we are saying we can potentially re-encrypt secrets that were already created on a new data key, and that's okay?

Reply:
It won't cause any problems afaik. If it's impractical to fix, how about writing a unit test that checks this scenario?

Agree this needs to be addressed longer term, so we can leave a TODO for now. Let's have more offline discussions about this 😄